### PR TITLE
Fix: #254 by increasing timeout, use backoff module instead of flaky

### LINKED ISSUE #254: translate_v3_batch_translate_text_with_glossary_test: test_batch_translate_text_with_glossary failed
Note: #191 was also for this test, but it was closed more than 10 days ago. So, I didn't mark it flaky.

----

commit: aa38acb3ebc771504887082f34fcc460951d910e
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/3c501be4-3d5d-4696-8334-57f90395b72c), [Sponge](http://sponge2/3c501be4-3d5d-4696-8334-57f90395b72c)
status: failed
<details><summary>Test output</summary><br><pre>target = functools.partial(<bound method PollingFuture._done_or_raise of <google.api_core.operation.Operation object at 0x7fd82823c9d0>>)
predicate = <function if_exception_type.<locals>.if_exception_type_predicate at 0x7fd8285cfb80>
sleep_generator = <generator object exponential_sleep_generator at 0x7fd8281a0c80>
deadline = 320, on_error = None

    def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
        """Call a function and retry if it fails.
    
        This is the lowest-level retry helper. Generally, you'll use the
        higher-level retry helper :class:`Retry`.
    
        Args:
            target(Callable): The function to call and retry. This must be a
                nullary function - apply arguments with `functools.partial`.
            predicate (Callable[Exception]): A callable used to determine if an
                exception raised by the target should be considered retryable.
                It should return True to retry or False otherwise.
            sleep_generator (Iterable[float]): An infinite iterator that determines
                how long to sleep between retries.
            deadline (float): How long to keep retrying the target. The last sleep
                period is shortened as necessary, so that the last retry runs at
                ``deadline`` (and not considerably beyond it).
            on_error (Callable[Exception]): A function to call while processing a
                retryable exception.  Any error raised by this function will *not*
                be caught.
    
        Returns:
            Any: the return value of the target function.
    
        Raises:
            google.api_core.RetryError: If the deadline is exceeded while retrying.
            ValueError: If the sleep generator stops yielding values.
            Exception: If the target raises a method that isn't retryable.
        """
        if deadline is not None:
            deadline_datetime = datetime_helpers.utcnow() + datetime.timedelta(
                seconds=deadline
            )
        else:
            deadline_datetime = None
    
        last_exc = None
    
        for sleep in sleep_generator:
            try:
>               return target()

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py:190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <google.api_core.operation.Operation object at 0x7fd82823c9d0>
retry = <google.api_core.retry.Retry object at 0x7fd8285d09d0>

    def _done_or_raise(self, retry=DEFAULT_RETRY):
        """Check if the future is done and raise if it's not."""
        kwargs = {} if retry is DEFAULT_RETRY else {"retry": retry}
    
        if not self.done(**kwargs):
>           raise _OperationNotComplete()
E           google.api_core.future.polling._OperationNotComplete

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/future/polling.py:89: _OperationNotComplete

The above exception was the direct cause of the following exception:

self = <google.api_core.operation.Operation object at 0x7fd82823c9d0>
timeout = 320, retry = <google.api_core.retry.Retry object at 0x7fd8285d09d0>

    def _blocking_poll(self, timeout=None, retry=DEFAULT_RETRY):
        """Poll and wait for the Future to be resolved.
    
        Args:
            timeout (int):
                How long (in seconds) to wait for the operation to complete.
                If None, wait indefinitely.
        """
        if self._result_set:
            return
    
        retry_ = self._retry.with_deadline(timeout)
    
        try:
            kwargs = {} if retry is DEFAULT_RETRY else {"retry": retry}
>           retry_(self._done_or_raise)(**kwargs)

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/future/polling.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (), kwargs = {}
target = functools.partial(<bound method PollingFuture._done_or_raise of <google.api_core.operation.Operation object at 0x7fd82823c9d0>>)
sleep_generator = <generator object exponential_sleep_generator at 0x7fd8281a0c80>

    @functools.wraps(func)
    def retry_wrapped_func(*args, **kwargs):
        """A wrapper that calls target function with retry."""
        target = functools.partial(func, *args, **kwargs)
        sleep_generator = exponential_sleep_generator(
            self._initial, self._maximum, multiplier=self._multiplier
        )
>       return retry_target(
            target,
            self._predicate,
            sleep_generator,
            self._deadline,
            on_error=on_error,
        )

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py:283: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

target = functools.partial(<bound method PollingFuture._done_or_raise of <google.api_core.operation.Operation object at 0x7fd82823c9d0>>)
predicate = <function if_exception_type.<locals>.if_exception_type_predicate at 0x7fd8285cfb80>
sleep_generator = <generator object exponential_sleep_generator at 0x7fd8281a0c80>
deadline = 320, on_error = None

    def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
        """Call a function and retry if it fails.
    
        This is the lowest-level retry helper. Generally, you'll use the
        higher-level retry helper :class:`Retry`.
    
        Args:
            target(Callable): The function to call and retry. This must be a
                nullary function - apply arguments with `functools.partial`.
            predicate (Callable[Exception]): A callable used to determine if an
                exception raised by the target should be considered retryable.
                It should return True to retry or False otherwise.
            sleep_generator (Iterable[float]): An infinite iterator that determines
                how long to sleep between retries.
            deadline (float): How long to keep retrying the target. The last sleep
                period is shortened as necessary, so that the last retry runs at
                ``deadline`` (and not considerably beyond it).
            on_error (Callable[Exception]): A function to call while processing a
                retryable exception.  Any error raised by this function will *not*
                be caught.
    
        Returns:
            Any: the return value of the target function.
    
        Raises:
            google.api_core.RetryError: If the deadline is exceeded while retrying.
            ValueError: If the sleep generator stops yielding values.
            Exception: If the target raises a method that isn't retryable.
        """
        if deadline is not None:
            deadline_datetime = datetime_helpers.utcnow() + datetime.timedelta(
                seconds=deadline
            )
        else:
            deadline_datetime = None
    
        last_exc = None
    
        for sleep in sleep_generator:
            try:
                return target()
    
            # pylint: disable=broad-except
            # This function explicitly must deal with broad exceptions.
            except Exception as exc:
                if not predicate(exc):
                    raise
                last_exc = exc
                if on_error is not None:
                    on_error(exc)
    
            now = datetime_helpers.utcnow()
    
            if deadline_datetime is not None:
                if deadline_datetime <= now:
>                   raise exceptions.RetryError(
                        "Deadline of {:.1f}s exceeded while calling {}".format(
                            deadline, target
                        ),
                        last_exc,
                    ) from last_exc
E                   google.api_core.exceptions.RetryError: Deadline of 320.0s exceeded while calling functools.partial(<bound method PollingFuture._done_or_raise of <google.api_core.operation.Operation object at 0x7fd82823c9d0>>), last exception:

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py:205: RetryError

During handling of the above exception, another exception occurred:

capsys = <_pytest.capture.CaptureFixture object at 0x7fd828215520>
bucket = <Bucket: tmp-e6db5a7fc03e4d96ae6166d29040bbab>

    @pytest.mark.flaky(max_runs=3, min_passes=1)
    def test_batch_translate_text_with_glossary(capsys, bucket):
>       translate_v3_batch_translate_text_with_glossary.batch_translate_text_with_glossary(
            "gs://cloud-samples-data/translation/text_with_glossary.txt",
            "gs://{}/translation/BATCH_TRANSLATION_GLOS_OUTPUT/".format(bucket.name),
            PROJECT_ID,
            GLOSSARY_ID,
            320,
        )

translate_v3_batch_translate_text_with_glossary_test.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
translate_v3_batch_translate_text_with_glossary.py:69: in batch_translate_text_with_glossary
    response = operation.result(timeout)
.nox/py-3-8/lib/python3.8/site-packages/google/api_core/future/polling.py:132: in result
    self._blocking_poll(timeout=timeout, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <google.api_core.operation.Operation object at 0x7fd82823c9d0>
timeout = 320, retry = <google.api_core.retry.Retry object at 0x7fd8285d09d0>

    def _blocking_poll(self, timeout=None, retry=DEFAULT_RETRY):
        """Poll and wait for the Future to be resolved.
    
        Args:
            timeout (int):
                How long (in seconds) to wait for the operation to complete.
                If None, wait indefinitely.
        """
        if self._result_set:
            return
    
        retry_ = self._retry.with_deadline(timeout)
    
        try:
            kwargs = {} if retry is DEFAULT_RETRY else {"retry": retry}
            retry_(self._done_or_raise)(**kwargs)
        except exceptions.RetryError:
>           raise concurrent.futures.TimeoutError(
                "Operation did not complete within the designated " "timeout."
            )
E           concurrent.futures._base.TimeoutError: Operation did not complete within the designated timeout.

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/future/polling.py:112: TimeoutError</pre></details>

### PR DESCRIPTION
**Pull request by nicain**
Here I have moved away from "flaky" in favor of "backoff", which was already a dependency of the library.

This repository currently uses the "flaky" module/plugin to pytest to allow retries on tests. As I was trying to quantify this flakiness, I quickly realized that under-the-hood, this module uses runtime modification (monkey-patching) of the usual pytest callbacks, which breaks the interface contracts that the pytest ecosystem subscribes to and does not let me aggregate the timing statistics I need for quantifying flakiness.

The "backoff" module is already in this repo. In this PR I modify the "usual" way we set up fixtures to use the callback system from "backoff". 

Fixes #254 by increasing timeout via backoff with 4 tries: [100, 250, 300, 500]

The worst execution time was 234 seconds, so the max allowed is now 500:
Flakiness report:
```
SUCCESS
       when    duration outcome  trial
0     setup    1.524859  passed      0
1      call  141.551338  passed      0
2  teardown    0.000545  passed      0
       when    duration outcome  trial
0     setup    1.418313  passed      1
1      call  234.604837  passed      1
2  teardown    0.000532  passed      1
       when    duration outcome  trial
0     setup    1.519378  passed      2
1      call  124.333620  passed      2
2  teardown    0.000748  passed      2
       when    duration outcome  trial
0     setup    1.569923  passed      3
1      call  129.542894  passed      3
2  teardown    0.000723  passed      3
       when    duration outcome  trial
0     setup    1.459673  passed      4
1      call  125.115383  passed      4
2  teardown    0.000780  passed      4
       when   duration outcome  trial
0     setup   1.515023  passed      5
1      call  79.950361  passed      5
2  teardown   0.000626  passed      5
       when   duration outcome  trial
0     setup   1.526374  passed      6
1      call  82.278016  passed      6
2  teardown   0.001218  passed      6
       when    duration outcome  trial
0     setup    1.533906  passed      7
1      call  107.097216  passed      7
2  teardown    0.000794  passed      7
       when    duration outcome  trial
0     setup    1.417850  passed      8
1      call  136.093182  passed      8
2  teardown    0.000509  passed      8
       when    duration outcome  trial
0     setup    1.433857  passed      9
1      call  134.896833  passed      9
2  teardown    0.000841  passed      9
```

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -11,4 +11,4 @@
 /samples/**/*.py                       @googleapis/cdpe-cloudai
 
 # The python-samples-owners team is the default owner for samples
-/samples/**/*.py                       @telpirion @sirtorry @googleapis/python-samples-owners
\ No newline at end of file
+/samples/**/*.py                       @aribray @sirtorry @googleapis/python-samples-owners
```
Comment (leahecole) [inline .github/CODEOWNERS:14]:
```suggestion
/samples/**/*.py                       @aribray @googleapis/python-samples-owners
```

File: samples/snippets/translate_v3_batch_translate_text_with_glossary_test.py [hunk 2]
```diff
@@ -37,15 +37,26 @@ def bucket():
...
 def test_batch_translate_text_with_glossary(capsys, bucket):
+
     translate_v3_batch_translate_text_with_glossary.batch_translate_text_with_glossary(
         "gs://cloud-samples-data/translation/text_with_glossary.txt",
         "gs://{}/translation/BATCH_TRANSLATION_GLOS_OUTPUT/".format(bucket.name),
         PROJECT_ID,
         GLOSSARY_ID,
-        320,
-    )
+        500)
```
Comment (leahecole) [inline samples/snippets/translate_v3_batch_translate_text_with_glossary_test.py:59]: per your suggestion change the 500 to a "max allowed time" parameter

**Review by leahecole — CHANGES_REQUESTED** [review]

File: samples/snippets/translate_v3_batch_translate_text_with_glossary_test.py [hunk 3]
```diff
@@ -37,15 +37,26 @@ def bucket():
...
+def bucket():
+    """Create a bucket feature for testing"""
+    return next(get_ephemeral_bucket())
+
+
+def on_backoff(invocation_dict):
+    """Backoff callback; create a testing bucket for each backoff run"""
+    invocation_dict['kwargs']['bucket'] = next(get_ephemeral_bucket())
+
+
```
[the remark this answers is not in the document]
Comment (nicain) [reply samples/snippets/translate_v3_batch_translate_text_with_glossary_test.py:50]: Done.

File: samples/snippets/translate_v3_batch_translate_text_with_glossary_test.py (hunk 2 above)
Comment (nicain) [reply samples/snippets/translate_v3_batch_translate_text_with_glossary_test.py]: Done.

**Review by aribray — APPROVED** [review]
Looks good to me!

**Review by leahecole — APPROVED** [review]

**Pull request merged**